apps/src/lib/proto/types.rs

```diff
@@ -0,0 +1,457 @@
+}
+
+pub enum RpcMessage {
+    IntentMessage(IntentMessage),
+    SubscribeTopicMessage(SubscribeTopicMessage),
+    Dkg(Dkg),
+}
+
+impl From<RpcMessage> for services::RpcMessage {
+    fn from(message: RpcMessage) -> Self {
+        let message = match message {
+            RpcMessage::IntentMessage(m) => {
+                services::rpc_message::Message::Intent(m.into())
+            }
+            RpcMessage::SubscribeTopicMessage(m) => {
+                services::rpc_message::Message::Topic(m.into())
+            }
+            RpcMessage::Dkg(d) => services::rpc_message::Message::Dkg(d.into()),
+        };
+        services::RpcMessage {
+            message: Some(message),
+        }
+    }
+}
+
+impl RpcMessage {
+    pub fn new_intent(intent: Intent, topic: String) -> Self {
+        RpcMessage::IntentMessage(IntentMessage::new(intent, topic))
+    }
+
+    pub fn new_topic(topic: String) -> Self {
+        RpcMessage::SubscribeTopicMessage(SubscribeTopicMessage::new(topic))
+    }
+
+    pub fn new_dkg(dkg: Dkg) -> Self {
+        RpcMessage::Dkg(dkg)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct IntentMessage {
+    pub intent: Intent,
+    pub topic: String,
+}
+
+impl TryFrom<services::IntentMessage> for IntentMessage {
+    type Error = Error;
+
+    fn try_from(message: services::IntentMessage) -> Result<Self> {
+        match message.intent {
+            Some(intent) => Ok(IntentMessage {
+                intent: intent.into(),
+                topic: message.topic,
+            }),
+            None => Err(Error::NoIntentError),
+        }
+    }
+}
+
+impl From<IntentMessage> for services::IntentMessage {
+    fn from(message: IntentMessage) -> Self {
+        services::IntentMessage {
+            intent: Some(message.intent.into()),
+            topic: message.topic,
+        }
+    }
+}
+
+impl IntentMessage {
+    pub fn new(intent: Intent, topic: String) -> Self {
+        IntentMessage { intent, topic }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SubscribeTopicMessage {
+    pub topic: String,
+}
+
+impl From<services::SubscribeTopicMessage> for SubscribeTopicMessage {
+    fn from(message: services::SubscribeTopicMessage) -> Self {
+        SubscribeTopicMessage {
+            topic: message.topic,
+        }
+    }
+}
+
+impl From<SubscribeTopicMessage> for services::SubscribeTopicMessage {
+    fn from(message: SubscribeTopicMessage) -> Self {
+        services::SubscribeTopicMessage {
+            topic: message.topic,
+        }
+    }
+}
+
+impl SubscribeTopicMessage {
+    pub fn new(topic: String) -> Self {
+        SubscribeTopicMessage { topic }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Intent {
+    pub data: Vec<u8>,
+    pub timestamp: Timestamp,
+}
+
+impl From<types::Intent> for Intent {
+    fn from(intent: types::Intent) -> Self {
+        let timestamp = match intent.timestamp {
+            Some(t) => t,
+            None => std::time::SystemTime::now().into(),
```

Comment:
I think same in here as for `Tx`, the timestamp should be required. Is that right, @sribst ?